File: Linked_list/linked_list.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class Linked_list {
    constructor() {
        this.head = null;
    }

    insertAtBeginning(data) {
        const newNode = new Node(data);
        newNode.next = this.head;
        this.head = newNode;
    }

    printList() {
        if (!this.head) {
            console.log('Linked List is empty');
            return;
        }

        let itr = this.head;
        let listStr = '';
        while (itr) {
            listStr += `${itr.data}-->`;
            itr = itr.next;
        }
        console.log(listStr);
    }

    insertAtEnd(data) {
        const newNode = new Node(data);
        if (!this.head) {
            this.head = newNode;
            return;
        }
        let itr = this.head;
        while (itr.next) {
            itr = itr.next;
        }
        itr.next = newNode;
    }

    insertValue(dataList) {
        this.head = null;
        for (const data of dataList) {
            this.insertAtEnd(data);
        }
    }

    getLength() {
        let count = 0;
        let itr = this.head;
        while (itr) {
            count++;
            itr = itr.next;
        }
        return count;
    }

    removeAt(index) {
        if (index < 0 || index >= this.getLength()) {
            throw new Error("Invalid index");
        }

        if (index === 0) {
            this.head = this.head.next;
            return;
        }

        let count = 0;
        let itr = this.head;
        while (itr) {
            if (count === index - 1) {
                itr.next = itr.next.next;
                break;
            }
            count++;
            itr = itr.next;
        }
    }

    insertAt(index, data) {
        if (index < 0 || index > this.getLength()) {
            throw new Error("Invalid index");
        }

        if (index === 0) {
            this.insertAtBeginning(data);
            return;
        }

        let count = 0;
        let itr = this.head;
        while (itr) {
            if (count === index - 1) {
                const newNode = new Node(data);
                newNode.next = itr.next;
                itr.next = newNode;
                break;
            }
            count++;
            itr = itr.next;
        }
    }
}

// Example usage:
const li = new Linked_list();
li.insertAtBeginning(1);
li.insertAtBeginning(2);
li.insertAtBeginning(3);
li.insertAtBeginning(4);
li.insertAtEnd(9);
li.insertAtBeginning(0);
li.printList();

const li2 = new Linked_list();
const dataList = ['Orange', 'water', 'cocaca', 'banana'];
li2.insertValue(dataList);
li2.printList();
console.log("li2 length:", li2.getLength());
li2.removeAt(2);
console.log("li2 after remove node");
li2.printList();
console.log("li2 after insert new value to a specific index");
li2.insertAt(1, "cheese");
li2.printList();

File: Linked_list/linked-list.go
package main

import (
	"fmt"
)

type Node struct {
	value interface{}
	next  *Node
}

type LinkedList struct {
	head *Node
}

func NewNode(value interface{}, next *Node) *Node {
	return &Node{value: value, next: next}
}

func NewLinkedList() *LinkedList {
	return &LinkedList{head: nil}
}

func (ll *LinkedList) InsertAtBeginning(data interface{}) {
	node := NewNode(data, ll.head)
	ll.head = node
}

func (ll *LinkedList) PrintList() {
	if ll.head == nil {
		fmt.Println("Linked List is empty")
		return
	}

	itr := ll.head
	listStr := ""
	for itr != nil {
		listStr += fmt.Sprintf("%v--", itr.value)
		itr = itr.next
	}
	fmt.Println(listStr)
}

func (ll *LinkedList) InsertAtEnd(data interface{}) {
	if ll.head == nil {
		ll.head = NewNode(data, nil)
		return
	}

	itr := ll.head
	for itr.next != nil {
		itr = itr.next
	}

	itr.next = NewNode(data, nil)
}

func (ll *LinkedList) InsertValues(dataList []interface{}) {
	ll.head = nil
	for _, data := range dataList {
		ll.InsertAtEnd(data)
	}
}

func (ll *LinkedList) GetLength() int {
	count := 0
	itr := ll.head
	for itr != nil {
		count++
		itr = itr.next
	}
	return count
}

func (ll *LinkedList) RemoveAt(index int) {
	if index < 0 || index >= ll.GetLength() {
		panic("Invalid value of index")
	}

	if index == 0 {
		ll.head = ll.head.next
		return
	}

	count := 0
	itr := ll.head
	for itr != nil {
		if count == index-1 {
			itr.next = itr.next.next
			break
		}
		count++
		itr = itr.next
	}
}

func (ll *LinkedList) InsertAt(index int, data interface{}) {
	if index < 0 || index > ll.GetLength() {
		panic("Invalid value of index")
	}

	if index == 0 {
		ll.InsertAtBeginning(data)
		return
	}

	count := 0
	itr := ll.head
	for itr != nil {
		if count == index-1 {
			node := NewNode(data, itr.next)
			itr.next = node
			return
		}
		count++
		itr = itr.next
	}
}

func main() {
	li := NewLinkedList()
	li.InsertAtBeginning(1)
	li.InsertAtBeginning(2)
	li.InsertAtBeginning(3)
	li.InsertAtBeginning(4)
	li.InsertAtEnd(9)
	li.InsertAtBeginning(0)
	li.PrintList()

	li2 := NewLinkedList()
	dataList := []interface{}{"Orange", "water", "cocaca", "banana"}
	li2.InsertValues(dataList)
	li2.PrintList()
	fmt.Println("li2 length:", li2.GetLength())
	li2.RemoveAt(2)
	fmt.Println("li2 after remove node")
	li2.PrintList()
	fmt.Println("li2 after insert new value to a specific index")
	li2.InsertAt(1, "cheese")
	li2.PrintList()
}
